### PR TITLE
refactor(gatsby): consolidate data layer

### PR DESCRIPTION
**Pull request by vladar**
## Description

This PR is a follow-up for #31371. It doesn't introduce any changes to behavior. It is simply a clean-up - moved various pieces of the data layer to where they belong for higher cohesion.

### CODE REVIEW

File: packages/gatsby/src/datastore/in-memory/indexing.ts [hunk 1]
```diff
@@ -63,109 +62,6 @@ export interface IFilterCache {
...
-/**
- * Get all nodes of type from redux store.
- */
-export const getNodesByType = (type: string): Array<IGatsbyNode> =>
-  getDataStore().getNodesByType(type)
-
-/**
- * Get all type names from redux store.
- */
-export const getTypes = (): Array<string> => getDataStore().getTypes()
```
Comment (vladar) [inline packages/gatsby/src/datastore/in-memory/indexing.ts:86]: Moved those convenience methods to `datastore/index.ts`

File: packages/gatsby/src/datastore/in-memory/indexing.ts [hunk 2]
```diff
@@ -63,109 +62,6 @@ export interface IFilterCache {
...
-  if (!node) {
-    console.error(
-      `getNodeAndSavePathDependency failed for node id: ${id} as it was not found in cache`
-    )
-    return undefined
-  }
-
-  createPageDependency({ path, nodeId: id })
-  return node
-}
```
Comment (vladar) [inline packages/gatsby/src/datastore/in-memory/indexing.ts:118]: Moved both those functions to `utils/nodes.ts`

File: packages/gatsby/src/datastore/in-memory/indexing.ts [hunk 3]
```diff
@@ -63,109 +62,6 @@ export interface IFilterCache {
...
-    }
-    store.dispatch({
-      type: `SET_RESOLVED_NODES`,
-      payload: {
-        key: typeName,
-        nodes: resolvedNodes,
-      },
-    })
-  }
-}
```
Comment (vladar) [inline packages/gatsby/src/datastore/in-memory/indexing.ts:143]: Moved this method to `schema/node-model.js` (where the only usage of this method is)

File: packages/gatsby/src/datastore/in-memory/indexing.ts [hunk 4]
```diff
@@ -63,109 +62,6 @@ export interface IFilterCache {
...
-  }
-
-  const resolvedNodes = resolvedNodesCache.get(typeName)
-
-  if (resolvedNodes) {
-    node.__gatsby_resolved = resolvedNodes.get(id)
-  }
-
-  return node
-}
```
Comment (vladar) [inline packages/gatsby/src/datastore/in-memory/indexing.ts:167]: This is no longer used anywhere, so removed

File: packages/gatsby/src/redux/save-state.js [hunk 5]
```diff
@@ -16,18 +15,7 @@ async function saveState() {
...
-const saveStateDebounced = _.debounce(saveState, 1000)
-
-/**
- * Starts listening to redux actions and triggers a database save to
- * disk upon any action (debounced to every 1 second)
- */
-function startAutosave() {
-  saveStateDebounced()
-  emitter.on(`*`, () => saveStateDebounced())
-}
```
Comment (vladar) [inline packages/gatsby/src/redux/save-state.js:28]: No longer used anywhere, so removed.

**Review by pieh — APPROVED** [review]
A lot cleaner. Thank you!

**Pull request merged**